### ISSUE TITLE
Automate EKS Cluster Deployment and Tear Down (v.next)

### ISSUE THREAD

**Issue by raywangoctova** [department-of-veterans-affairs/va.gov-team#50671]
## Status
| Date | Status | Launch Date (see above) | Notes |
| ----- | ------ | ------------------------- | ------ |
|2025/12/05 | In-Progress | End of 2026 | With the Gov. shutdown lifted, we will merge a PR this Sprint (16 OY3) that frees up a nearly 90 IPs that will provide us the freedom/ability to implement v.Next into our Prod tier.  We have also implemented Taragon which allows auditing at the Pod level, which has never been done until now.  We have one more more ticket, due in Sprint 17 (OY3), to create an ArgoCD app. of app's in the development environment which then allow us to begin working with application teams to on-board to v.Next dev. |
| 2025/10/29 | Blocked | End of 2026 | We are still waiting for the Gov. shutdown to end so we can merge the code for adding subnets, making available unused IP addresses.  This will then allow us to release v.Next into our prod environment.  In addition, we had a meeting last week to identify the additional v.Next work necessary to begin on-boarding applications into the v.Next dev tier to begin testing. |
| 2025/10/16 | Blocked | TBD | v.Next has been deployed to Dev., Stagging and Sandbox.  We have run out of IP addresses in Prod. to continue with our work deploying to that environment.  We need to complete the work to add subnet so Prod. to unlock unused up IPs.  The Gov. Shut-Down is blocking us from completing this work. |
| 6/20/25 | In-Progress | On-Track Q3 | Argo CD Rollouts completed, Velero Helm Chart integrated, beginning to work on helm chart updates in the coming sprint. |
| 5/30/25 | In-Progress | On-Track Q3 | Updating Argo CD, ensure images are being pulled from ECR, setup ASG per availability zone, cump helm deploy timeout, fix port conflict, use S3 locking instead of DynamoDB |
| 5/16/25 | In-Progress | On-Track | Enhancing vNext cluster reliability through improved automation - focused on migrating remaining images to ECR, completing SQS setup for Node Termination Handler, refining ASG Warm Pool, and implementing Default Network Policy for EKS environments. |
| 5/1/25 | In-Progress | On-Track | improved vNext cluster automation by fixing Kyverno deployment dependencies to prevent timeouts, improving subnet handling across VPC configurations, and implementing high availability with triple redundancy for all Kyverno controllers to enhance system reliability. |
| 4/18/25 | In-Progress | On-Track | No update this sprint as we focused on Preview Environments |
| 4/3/25 | In-Progress | On-Track | Working to set up Amazon SQS (Simple Queue Service) for the AWS Node Termination Handler in the EKS cluster environment. |
| 3/21/25 | In-Progress | On-Track | Enhanced infrastructure by implementing zone-specific ASGs, deploying optimal node distribution by deploying tooling through Velero and Kyverno helm charts, improving our infra as code maturity and posture. |
| 3/7/25 | In-Progress | On-Track | Enhanced EKS cluster architecture by implementing one ASG per Availability Zone for improved resilience. Transferred Velero Helm chart backup responsibilities. Tested end-to-end mTLS with our service mesh. Evaluating Gateway API as an alternative to Ingress for managing inbound traffic while keeping all services current with updates  |
| 2/24/25 | In-Progress | On-Track | Implemented self-managed ASG for improved availability zone resilience, expanded storage class options, and deployed ASG warm pools to reduce scaling response times. |
| 2/10/25 | In-Progress | On-Track | Team is reviewing an upgrade to v.Next clusters 1.32 within EKS. |
| 1/20/25 | In-Progress | On-Track | EKS Utility Cluster Upgrade planned for 1/23/24 at 11pm ET. |
| 1/6/25 | In-Progress | On-Track | No new updates. |
| 12/9/24 | In-Progress | On-Track | v.Next clusters continue to be available. Team is fine tuning them for better long-term maintenance. A migration strategy is still needed. |
| 11/25/24 | In-Progress | On-Track | No new updates. |
| 10/28/24 | In-Progress | On-Track | Team is still reviewing EKS1.31 upgrade, which introduces some changes that we need to account for. |
| 10/21/24 | In-Progress | On-Track | Argo authentication updates and reviewing changes needed for EKS 1.31 (latest EKS release). |
| 10/14/24 | In-Progress | On-Track | Breaking out authentication and improving scaling operations (lowering time it takes for applications and the cluster to scale). |
| 10/7/24 | In-Progress | On-Track | Continuing to move everything into ECR as well as constant small improvements. |
| 9/30/24 | In-Progress | On-Track | Moving the remaining couple of images into ECR. Also modifying warm pool and over provisioning for faster scaling operations. |
| 9/23/24 | In-Progress | On-Track | v.Next clusters are getting continuous upgrades. |
| 9/16/24 | In-Progress | On-Track | Enabled the ability to not expose Argo, per Security request. |
| 9/9/24 | In-Progress | On-Track | Team is testing a new Argo change to prevent the Argo UI from getting deployed for Production clusters to help with Security concerns regarding prod access via SOCKS. This is for new clusters, and new Argo, not existing. |
| 9/3/24 | In-Progress | On-Track | Moving to protect Argo in Production based on suggestions. Have deployed and torn down hundreds of clusters thus far. |
| 8/26/24 | In-Progress | On-Track | With the new IPs provided by VAEC last week, the team has added them to DEV. Currently using the new 100.x CIDR blocks for our new (Platform v.Next) clusters. |
| 8/19/24 | In-Progress | On-Track | Team had a meeting with VAEC, they are going to give us 3 100.x/24 CIDRs for our dev VPC. This should be enough to unblock our work, and we can adjust from there if more are needed. |
| 8/6/24 | In-Progress | On-Track | Reviewing the private IPs and working through what is needed for the VPCs and the existing PPCs with Reliability. |

## Problem Statement
Currently the EKS clusters that host `vets-api` and other applications are provisioned in manual, time intensive ways. This creates overhead when Platform teams are performing updates, maintenance, and in the overall management of these clusters. These tasks include:
- Creating new environments (both live and ephemeral)
- Performing blue/green upgrades to EKS clusters
- Monitoring and logging those clusters

In addition, developers do not have the ability to spin up and tear down clusters without risk of creating orphaned resources.

Our hypothesis is that we can solve these problems by implementing improved automation for the deployment, management, and decommissioning of these EKS clusters.

This would support the goals of zero downtime deploys using blue/green deployment, and facilitate the ongoing maintenance and updates of EKS control plane.

## User Impact

- All teams that are building, maintaining, or deploying a Kubernetes (EKS) cluster on Platform
- In the event of an upgrade gone wrong, we have the previous cluster still running to [DNS] switch back to
  - Creates a more stable and reliable platform for our end users and engineers
- Provides a safe environment for building and testing new core Platform services (Ingress, Networking, etc.)

## Where was this problem reported?

- Going through multiple in-place cluster upgrades where rolling back is not an option
- Disaster recovery exercises
- Developers and system users looking for a safe place to learn and work with the Platform's cluster setup

## What do we not know about the problem space?

*What areas do we need more information about so that we can accurately solve the problem? What discovery is needed?*

- Using the 172.16.0.0/16 CIDR for cluster nodes
  - Platform's 10.0.0.0/8 CIDR is incredibly limited (in-place cluster upgrades had to wait for IPs at times)
  - Only the cluster load balancers require VA resolution on the 10.0.0.0/8 CIDR
- Outside of when the current EKS clusters were built, no other cluster has been built and deployed with our current set of tools and applications

## What (if any) research or discovery has been done?

This has been a topic since mid to late 2022, and has undergone extensive research into use cases that today we cannot accommodate with our existing infrastructure

- https://vfs.atlassian.net/wiki/spaces/OT/pages/2522448886/Platform+v.+Next+Goals
- https://vfs.atlassian.net/wiki/spaces/OT/pages/2333802651/Kubernetes+Production-Ready+Checklist+EKS
- [https://vfs.atlassian.net/wiki/spaces/OT/pages/1549041761/Infrastructure+Roadmap+WIP#Future-initiatives-(in-logical-priority)](https://vfs.atlassian.net/wiki/spaces/OT/pages/1549041761/Infrastructure+Roadmap+WIP#Future-initiatives-%28in-logical-priority%29 "https://vfs.atlassian.net/wiki/spaces/OT/pages/1549041761/Infrastructure+Roadmap+WIP#Future-initiatives-(in-logical-priority)")
- Through experience in maintaining the existing infrastructure

## What is the acceptance criteria?

- [x] EKS clusters are deployed using Terraform
  - [x] No resources are shared between clusters
- [x] Every cluster has it's own unique vfs.va.gov URL (i.e., same way that existing clusters have traefik-{{env}}.vfs.va.gov)
- [ ] Clusters are deployed and torn down through automation
	- [x] Allow Platform engineers to spin up and manage their own ephemeral EKS cluster
	- [x] Environments (live clusters)
	- [x] Ephemeral
		- [ ] Scheduled reminders to tear down ephemeral clusters
	- [ ] Nightly (fresh build and teardown)
	- [ ] Nightly Upgrade (fresh deploy + cluster upgrade + teardown)
        - [ ] Nightly Next (fresh deploy of the next version of EKS)
- [ ] Deploy smoke test with each cluster [to verify that it can be deployed to as well as serve traffic]
- [ ] Develop a tech stack upgrade schedule/roadmap
- [ ] Documentation (Update existing, and/or create new)
  - [ ] Internal cluster architecture diagrams
  - [ ] Steps for a Blue/Green EKS environment
  - [ ] Include steps for switching back and forth
- [x] The existing observability stack is deployed
- [ ] Successfully deploy EKS version 1.27 to the dev environment using A/B deployment

## How should we measure success?
Time to deploy a new EKS cluster
Time to teardown a new EKS cluster
Time to perform blue/green deploy of EKS cluster
of orphaned resources remaining after EKS cluster teardown
of manual actions required to update EKS cluster control plane version

**Comment by npeterson54** [comment]
Action items: 
- [ ] Clarify the user impact
- [ ] Success metrics (how do we know we are successful) 
- [ ] Make a decision on the last 2 AC.  If not, create new projects for them.

**Comment by npeterson54** [comment]
Random thoughts; mostly just reiterating what was said yesterday.  To avoid scope creep, I think we should break the last two AC out into their own project.  They make an awesome project and do not fit into the problem statement of the issue.

**Comment by EWashb** [comment]
@npeterson54 I'm reading through all the documentation so that I can understand the landscape here. I'm curious if yall have even more thought around the metrics we can track. I see above you mentioned # of failures per month, but would love to jam about your ideas!

**Comment by mchelen-gov** [comment]
Quick process note, @EWashb should be the only one adding the "PO Endorsed" label to this project. Erika as soon as you feel it is ready from a refinement perspective, please re-add. Thanks everyone for understanding!

**Comment by EWashb** [comment]
This looks good to me. Thank you @npeterson54 for getting those metrics defined!

**Comment by EWashb** [comment]
@little-oddball thanks for you comments. Everything is fine from my perspective to start this work. Since I am now onboarded, hopefully that will not be a blocker or a source of confusion again.

**Comment by annekerr49** [comment]
Erika is this is your ticket could you please add it to the DE Procuts -In progress column ?  https://github.com/orgs/department-of-veterans-affairs/projects/940/views/2

**Comment by EWashb** [comment]
@annekerr49 it has now been added to the DE board.

**Comment by EWashb** [comment]
@npeterson54 as we are completing tasks within this epic, can you be sure to check off the ACs listed so that I can follow along with the progress? Thank you.

**Comment by EWashb** [comment]
@AparnaNittalaUSDS going through our DE product board and I added the start date to this since I was still POing this team at that time. However, I don't know the target delivery date for this. Could you add it here?

**Comment by AparnaNittalaUSDS** [comment]
@annekerr49 Jan 2024 is the latest date for the completion of this activity. Confirmed with Nate

**Comment by AparnaNittalaUSDS** [comment]
Updating the target delivery date to February 28, 2025

**Comment by annekerr49** [comment]
What is the reason this project is targeting Feb 28, 2015 as end date?

**Comment by ph-One** [comment]
This work has been on hold for many months while all of Platform was being migrated to Amazon Linux 2 based images (Amazon Linux 1 is end of life). We are beginning to get back to this work.

**Comment by kflow85** [comment]
_Aug 6, 2024_ : Kyle M is back on this. Reviewing the private IPs and working through what is needed for the VPCs and the existing PPCs with Reliability. As of 1:35pm in meeting, we are out of IP space. Kyle M is point-of-contact to mediate this issue and is working with VAEC 

_Aug 19, 2024_ : Team had a meeting with VAEC, they are going to give us 3 100.x/24 CIDRs for our dev VPC. This should be enough to unblock our work, and we can adjust from there if more are needed.

_Aug 26, 2024_ : With the new IPs provided by VAEC last week, the team has added them to DEV. Currently using the new 100.x CIDR blocks for our new (Platform v.Next) clusters.

_Sep 3, 2024_ : Moving to protect Argo in Production based on suggestions. Have deployed and torn down hundreds of clusters thus far.

_Sep 9, 2024_ : Team is testing a new Argo change to prevent the Argo UI from getting deployed for Production clusters to help with Security concerns regarding prod access via SOCKS. This is for new clusters, and new Argo, not existing.

_Sep 16, 2024_ : Enabled the ability to not expose Argo, per Security request.

_Sep 23, 2024_ : v.Next clusters are getting continuous upgrades.

_Sep 30, 2024_ : Moving the remaining couple of images into ECR. Also modifying warm pool and over provisioning for faster scaling operations.

_Oct 7, 2024_ : Continuing to move everything into ECR as well as constant small improvements.

_Oct 14, 2024_ : Breaking out authentication and improving scaling operations (lowering time it takes for applications and the cluster to scale).

_Oct 21, 2024_ : Argo authentication updates and reviewing changes needed for EKS 1.31 (latest EKS release).

_Oct 28, 2024_ : Team is still reviewing EKS1.31 upgrade, which introduces some changes that we need to be accounted for.

**Comment by EWashb** [comment]
@JoeTice @jwoodman5 still on track?

**Comment by ph-One** [comment]
This body of work is for what we're referring to as "v.Next" clusters, and they will be what we deploy into VAEC. The work you've seen recently is merely to continue upgrading and maintaining existing systems to allow us the runway to focus on the VAEC migration.

When this all started we intended to cut over to v.Next in our existing AWS, but it didn't make sense to switch clusters twice. That said, we regularly deploy v.Next clusters for testing (such as the AMI that we will be rolling out with cluster upgrades tonight, and to identify potential helm chart changes needed for tooling upgrades); I believe the new Preview Environments will live in a v.Next cluster as well.

**Comment by EWashb** [comment]
@JoeTice @jwoodman5 given the note above, we need to rescope this work or at least indicate the relationship with VAEC.

**Comment by EWashb** [comment]
@davidpetras could you give this ticket a look over and see if it is still accurate for the current state of this work? How close are we to done?

**Comment by davidpetras** [comment]
@EWashb there is plenty of work(multiple sprints/months of work) that needs to be done and tickets created and attached to this effort, there will be some cross team dependencies with SRE as well. We will make sure this is updated after the upcoming holiday.

Thank you!

**Comment by EWashb** [comment]
Thanks @davidpetras. We are tracking this on the OCTO Product Board for our leadership, and the delivery date was 9/1. What do you think is a more reasonable date I can push that out to? (I have to report on this during a meeting this afternoon)

**Comment by EWashb** [comment]
I updated my target delivery date to a couple of months out to give us some time to understand security impacts

**Comment by EWashb** [comment]
Product outline: https://vfs.atlassian.net/wiki/spaces/TT2/pages/4601610241/EKS+Platform+v.Next
